File: Sliding Window/Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstringSet(self, s: str) -> int:
        
        window_start = 0

        longest = 0

        substring_chars = set()

        for window_end in range(len(s)):

            current_char = s[window_end]

            # we need to remove characters from the front of the window until we get rid of 
            # the first occurence of the current char, which is duplicated
            while current_char in substring_chars:

                front_char = s[window_start]

                substring_chars.remove(front_char)

                window_start += 1

            # add back in the current char, since we want to include it in the window
            substring_chars.add(current_char)

            longest = max(longest, len(substring_chars))

        return longest


    def lengthOfLongestSubstringDict(self, s: str) -> int:
        
        window_start = 0

        longest = 0

        # maps each character to its rightmost index that has been encountered
        substring_chars = {}

        for window_end in range(len(s)):

            current_char = s[window_end]

            if current_char in substring_chars:

                # window_start may already be ahead of the first occurence of the duplicate character.
                # if that's the case, we don't need to update window_start

                # if the first occurence of the duplicate character is after (or equal to) start, then we need to 
                # move start to the character after the first duplicate. 

                if substring_chars[current_char] >= window_start:
                    window_start = substring_chars[current_char] + 1                    
            
            # update the index of the current char to the current index
            substring_chars[current_char] = window_end
            
            longest = max(longest, window_end - window_start + 1)


        return longest

File: Old Java Solutions/Misc/Longest Substring Without Repeating Characters.java
class Solution {

	public int lengthOfLongestSubstring(String s) {
		int n = s.length();
		HashSet<Character> charSet = new HashSet<>();

		int result = 0;
		int i = 0;
		int j = 0;

		while (i < n && j < n) {
			char currentChar = s.charAt(j);
			if (!charSet.contains(currentChar)) {
				charSet.add(currentChar);
				j++;
				result = Math.max(result, j - i);

			} else {
				i ++;
				charSet.remove(s.charAt(i));
			}
		}

		return result;

	}

    public int lengthOfLongestSubstring(String s) {

    	int result = 0;

        for (int i = 0; i < s.length(); i ++) {

        	for (int j = i + 1; j <= s.length(); j ++) {
        		String sub = s.substring(i, j);
        		if (allUnique(sub)) {
        			result = Math.max(result, sub.length());
        		}
        	}


        }

        return result;
    }

    public boolean allUnique(String s) {

    	HashSet<Character> characters = new HashSet<>();

    	for (int i = 0; i < s.length(); i ++) {
    		char current = s.charAt(i);
    		if (characters.contains(current)) {
    			return false;
    		}

    		characters.add(current);

    	}

    	return true; 

    }
}